### ISSUE TITLE
Forgetting "dyn" before Fn() + Trait causes panic

### ISSUE THREAD

**Issue by thanatos** [dtolnay/async-trait#89]
In non-async-trait code, this is currently legal, though it emits a warning that it should have the `dyn` keyword:

```
impl<F> Foo for Fn(i8) -> F + Send + Sync
  where F: ... {}
```

In async-trait code, this causes a panic:
```
#[async_trait::async_trait]
impl<F> RequestHandler for Fn(&Request) -> F + Send + Sync
  where F: ... {}
```

```
error: custom attribute panicked
  --> src/lib.rs:15:1
   |
15 | #[async_trait::async_trait]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: message: unexpected token
```

[A full example in the Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=9807f69a0c79b48c21cafb6c9b84357f)

I'm not sure whether the code should be *accepted*, given that the current compiler warns that the `impl` w/o the `dyn` is deprecated, but I figure `async_trait` might want to either accept w/ a similar warning, or at the very least reject with a clearer error message.

(I mostly stumbled into this; I do not think the `impl` is anything _close_ to what I actually want in my real-world use case. The compiler's insistence on `dyn` is nice though, since it clued me in that I'm barking up the wrong syntactical tree.)

**Comment by dtolnay** [comment]
Minimized:

```rust
use async_trait::async_trait;

#[async_trait]
trait Trait {
    async fn f(&self);
}

#[async_trait]
impl Trait for Send + Sync {
    async fn f(&self) {}
}
```

**Comment by dtolnay** [comment]
Fixed on 0.1.30.

**Comment by thanatos** [comment]
Wow that was fast! Also, _thank you_ for an excellent library!
